src/data/representation.lean

```diff
@@ -0,0 +1,312 @@
+```
+(where the second function provides the `representation k G M` relative to the `module k M`
+coming from `restrict_scalars k (monoid_algebra k G) M`).
+These provide the equivalence between 1. and 2. above.
+
+The function `monoid_algebra.lift` in `data.monoid_algebra` then provides the equivalence
+between 3. and 5.
+
+(There's obviously a quite direct connection between points 2. and 5. above,
+which hasn't been developed yet.)
+
+(We don't yet say anything about point 4. above; perhaps it's less useful.)
+
+Finally, note that really all of these equivalences should be done as equivalences of categories,
+not just functions back and forth.
+We'll get there in a later file!
+
+## Tags
+
+linear action, representation
+-/
+
+universes u v w
+
+section prio
+set_option default_priority 100 -- see Note [default priority]
+/--
+A k-linear representation of a group G.
+-/
+@[nolint dangerous_instance] -- k is an `out_param`, so this is a false positive.
+class representation
+  (k : out_param $ Type v) [comm_ring k] (G : Type u) [monoid G] (M : Type w) [add_comm_group M] [module k M]
+  extends distrib_mul_action G M : Type (max u v w) :=
+(smul_smul : ∀ (g : G) (r : k) (m : M), g • (r • m) = r • (g • m))
+end prio
+
+section
+variables {k : Type v} {G : Type u} {M : Type w}
+variables [comm_ring k] [monoid G] [add_comm_group M] [module k M]
+
+/-- A group element acts by an k-linear map in any k-linear representation. -/
+def smul.linear_map [representation k G M] (g : G) : M →ₗ[k] M :=
+{ to_fun := λ m, g • m,
+  add := λ m m', by simp [smul_add],
+  smul := λ r m, by simp [representation.smul_smul], }
+
+@[simp]
+lemma smul.linear_map_apply [representation k G M] (g : G) (m : M) :
+  (smul.linear_map g : M →ₗ[k] M) m = g • m := rfl
+end
+
+namespace representation
+variables (k : Type v) (G : Type u)
+variables [comm_ring k] [monoid G]
+variables (M : Type w) [add_comm_group M] [module k M]
+
+/--
+A k-linear representation of G on M can be thought of as
+a multiplicative map from G into the k-linear endomorphisms of M.
+-/
+def as_monoid_hom [representation k G M] : G →* (M →ₗ[k] M) :=
+{ to_fun := λ g, smul.linear_map g,
+  map_one' := by { ext, simp, },
+  map_mul' := λ g g', by { ext, simp [mul_smul], }, }
+
+@[simp]
+lemma as_monoid_hom_apply_apply [representation k G M] (g : G) (m : M) :
+  (as_monoid_hom k G M) g m = g • m := rfl
+
+variables {k G M}
+
+/--
+We get a k-linear representation of G on M from
+a multiplicative map from G into the k-linear endomorphisms of M.
+-/
+def of_monoid_hom (ρ : G →* (M →ₗ[k] M)) : representation k G M :=
+{ smul := λ g m, ρ g m,
+  one_smul := λ m, by { simp, erw [monoid_hom.map_one ρ], simp, }, -- FIXME: anyone know why we need `erw`?
+  mul_smul := λ g g' m, by { simp, erw [monoid_hom.map_mul ρ], simp, }, -- FIXME: here to
+  smul_zero := λ g, by { simp, },
+  smul_add := λ g m m', by { simp, },
+  smul_smul := λ g r m, by { simp, }, }
```

Comment:
I see -- so I can use "rho". Then what is the point of `representation k G M`?

Reply:
To be honest, I'm pretty unconvinced by the usefulness of this definition. The main thing I was aiming for here was to make it easy to use `g • m` to talk about the actual action in terms of elements.

In a separate branch (not depending on this one) I've just proved Maschke's theorem, completely staying in the world of `module (monoid_algebra k G)`. As I discovered there, writing things out in terms of elements was not actually very useful! Much more useful was being able to talk about "multiplication by g as a k-linear map", i.e. exactly the `ρ : G →* (M →ₗ[k] M)` interpretation.

Reply:
I wonder if sometime soon we (you, me, anyone interested) could have a "meeting", whether just on zulip or with screen sharing, to go over the possibilities. I'd be happy to go through my proof of Maschke's theorem and explain what felt most painful there.

Reply:
Also --- I think the PR's this is stacked on top of #2417 and #2366, are worth reviewing separately from this one, and this one will be clearer and easier to see what it's doing once those have been looked at.

Reply:
> I wonder if sometime soon we (you, me, anyone interested) could have a "meeting", whether just on zulip or with screen sharing, to go over the possibilities. I'd be happy to go through my proof of Maschke's theorem and explain what felt most painful there.

@semorrison  I'm only seeing this comment now... but I think such a meeting is a great idea.

Reply:
I'm going to close this PR in the meantime. I've been meaning to write some notes about my other experiments so far. In any case, I don't think this PR is a particularly good approach.